### ISSUE TITLE
4/19 index selection

### ISSUE THREAD

**Issue by chenyan1114** [chenyan1114/WebSearchEngine#1]
## Overview
This document describes the implementation of  html keywords extraction, a tool for extracting keywords from crawled HTML pages using the YAKE (Yet Another Keyword Extractor) algorithm.
## Result Summary
Crawl the entire golden set (total 25,924 samples), perform keyword extraction on all crawled HTML pages, and check whether the extracted keywords match the queries.
- High coverage:
Using full text, ~75.7% of queries are matched within Top-10 and ~86.0% within Top-50 (any match), indicating that relevant keywords are usually extracted.
- Low exact match:
Exact@1 is only ~13.9%, much lower than substring/token matches (~40%)
- Title vs Full Text:
Title performs better in early ranking (higher MRR ~0.51, better Top-10)
Full text provides better overall coverage (higher Top-50)

## charts

<img width="1800" height="750" alt="Image" src="https://github.com/user-attachments/assets/f08dfd04-a248-4451-9dd6-14314c43558f" />

### Core Components
1. **HTML Text Extraction** (`extract_clean_text`)
   - Parses HTML using lxml
   - Extracts title from `<title>` tag
   - Extracts meta description from `<meta name="description">` or `<meta property="og:description">`
   - Removes script, style, noscript, svg, canvas, iframe, form, picture, source elements
   - Removes layout elements (nav, header, footer, aside)
   - Normalizes whitespace
2. **Language Detection** (`detect_language`)
   - Uses langdetect library
   - Falls back to "en" on detection failure
   - Can be set to fixed language via `--language` parameter
3. **Keyword Extraction** (YAKE)
   - Configurable n-gram size (1-3)
   - Deduplication threshold (0.0-1.0)
   - Top-K keywords per document
   - Window size for YAKE algorithm
### Dual Keyword Extraction
The tool extracts keywords from three text sources:

| Field | Description |
|-------|-------------|
| `keywords` | Full text (title + meta description + body) |
| `title_keywords` | Title only |
| `context_keywords` | Meta description + body |

This separation allows for better evaluation - sometimes title keywords perform better, sometimes context keywords, and combining them (taking minimum rank) can improve results.
### Example
```json
"query_keyword": "south africa vs india", 
"url": "https://www.youtube.com/watch?v=fVoNS00CqPE", "domain": "youtube.com",
"detected_language": "en",
"content_top_keywords": ["India v South Africa", "World Cup", "South Africa", "YouTube Watch highlights", "ICC Follow", "ICC Men", "Subscribe to our YouTube channel", "match of the ICC Men", "India v South", "Watch highlights"], 
"title_top_keywords": ["India v South Africa", "Match Highlights", "World Cup", "South Africa", "India v South", "Warm-ups", "YouTube", "Africa", "Match", "Highlights"], 
```
## Evaluation

The extracted keywords can be evaluated which computes:

- **Exact match**: Query equals keyword exactly
- **Substring match**: Query is substring of keyword or vice versa
- **Token cover**: All query tokens match tokens in keyword
- **MRR**: Mean Reciprocal Rank across top-K results

## Matching Methods
Crawl the entire golden set, perform keyword extraction on all crawled HTML pages, and check whether the extracted keywords match the queries.
### Exact Match

Query must equal keyword exactly (after compact normalization).

```python
def match_exact(query: str, keyword: str) -> bool:
    return query == keyword
```

Example:
```
query: "machine learning"
keyword: "machinelearning"  → False
keyword: "machine learning"   → True
```

### Substring Match

Query appears in keyword or keyword appears in query.

```python
def match_substring(query: str, keyword: str) -> bool:
    return query in keyword or keyword in query
```

Example:
```
query: "machine"
keyword: "machine learning" → True (query in keyword)
keyword: "deep learning"    → False
```

### Token Cover

All query tokens must be found in keyword tokens (exact, substring, or contained).

```python
def match_token_cover(query_tokens: list[str], keyword_tokens: list[str]) -> bool:
    for qt in query_tokens:
        if not any(qt == kt or qt in kt or kt in qt for kt in keyword_tokens):
            return False
    return True
```

Example:
```
query_tokens: ["machine", "learning"]
keyword_tokens: ["learning", "machine"] → True
keyword_tokens: ["machine"]              → False (missing "learning")
```

### Any Match

Any of exact/substring/token_cover matches.

```python
any_match = exact_match or substring_match or token_cover_match
```

### Top-K Hit Rate

Percentage of queries that have at least one match in top-K keywords.

```
hit_rate@k = (# queries with match in top K) / (total queries)
```

Example:
```
Top-1 hit rate: 5%
Top-3 hit rate: 10%
Top-10 hit rate: 20%
```

### MRR (Mean Reciprocal Rank)

Average of 1/rank for matched queries.

```
MRR = (1/rank_1 + 1/rank_2 + ... + 1/rank_n) / n
```

## charts

<img width="1500" height="900" alt="Image" src="https://github.com/user-attachments/assets/5bdc745f-9a24-439b-b343-aa3198e66b2c" />

<img width="2100" height="900" alt="Image" src="https://github.com/user-attachments/assets/fddcb805-cd2a-4e26-9072-09e03d022ef1" />

<img width="2100" height="900" alt="Image" src="https://github.com/user-attachments/assets/414ffb44-0b80-4119-a093-db37bce1d65a" />

## Limitations

1. Language detection may be inaccurate for short texts
2. Custom extractors must handle exceptions gracefully
3. HTML encoding issues may cause parsing failures
4. Very large HTML files may cause memory issues
5. Deduplication threshold may filter out valid keywords

### Custom Extractor Support

For domains with special HTML structure (e.g., YouTube), a custom extractor module can be provided:

```python
def extract(html_bytes: bytes) -> dict:
    """Extract title and description from domain-specific HTML."""
    return {
        "title": "Extracted Title",
        "description": "Extracted Description"
    }
```
#### Example
```python
# for Youtube.com
def extract(html_bytes):

    html = html_bytes.decode("utf-8", errors="ignore")
    match = re.search(r"var ytInitialData = ({.*?});", html)
    data = json.loads(match.group(1))
    title = data["contents"]["twoColumnWatchNextResults"]\
                ["results"]["results"]["contents"][0]\
                ["videoPrimaryInfoRenderer"]["title"]["runs"][0]["text"]
    desc_node = find_key(data, "attributedDescriptionBodyText")
    description = desc_node["content"]
    return {"title": title, "description": description}
```

### YAKE Configuration

- **Language**: auto-detected or fixed (e.g., "en", "es", "zh")
- **n-gram size**: 1-3 (unigrams, bigrams, trigrams)
- **Deduplication threshold**: 0.0-1.0 (higher = more aggressive deduplication)
- **Top-K**: Number of keywords to extract (default: 20)
- **Window size**: Context window for YAKE (default: 1)

### Text Preprocessing

1. Remove unwanted elements (script, style, noscript, etc.)
2. Remove layout elements (nav, header, footer, aside)
3. Extract title and meta description
4. Normalize whitespace (collapse multiple spaces to single space)

## Text Normalization

Before matching, both query and keywords are normalized:

### 1. Basic Normalization (`normalize_basic`)
- Unicode NFKC normalization
- Convert to lowercase (casefold)
- Keep only word characters (letters, numbers)
- Replace non-word chars with spaces
- Collapse whitespace

```
"Machine Learning" → "machine learning"
"COVID-19" → "covid 19"
```

### 2. Compact Normalization (`normalize_compact`)
- Unicode NFKC normalization
- Convert to lowercase
- Keep only word characters (no spaces)

```
"machine learning" → "machinelearning"
```

### 3. Tokenization (`tokenize`)
- Normalize using basic normalization
- Remove stopwords (a, an, the, for, etc.)
- Remove single ASCII characters
- Return list of tokens

```
"machine learning algorithm" → ["machine", "learning", "algorithm"]
```

**Comment by yangyoung0712** [comment]
# Index Selection (Batch3 vs Tuned Model)

## 📌 Baseline (Batch3 Index Coverage)

### Random Coverage
- URL: **1,443 / 7,978 = 18.09%**

### Head Coverage
- URL: **1,332 / 7,245 = 18.39%**

---

## 🚀 Tuned Model Result (`top_32m_419.tsv`)

### Random Coverage
- URL: **1,811 / 7,958 = 22.76%**

### Head Coverage
- URL: **1,372 / 7,436 = 18.45%**

---

## 📊 Improvement Summary

| Metric | Baseline | Tuned Model | Absolute Gain | Relative Gain |
|------|---------:|------------:|--------------:|--------------:|
| Random Coverage | 18.09% | 22.76% | +4.67  | +25.8% |
| Head Coverage   | 18.39% | 18.45% | +0.06  | +0.03% |

---

## 📈 Coverage Gain (URL Count)

- **Random**: +606 URLs (1,205 → 1,811)
- **Head**: +410 URLs (962 → 1,372)

---

## ⚙️ Model Configuration

### Original Default Parameters (`train_model.py`)
```python
learning_rate = 0.05
num_leaves = 63
num_boost_round = 500
early_stopping_rounds = 30
feature_fraction = 0.9
bagging_fraction = 0.8
bagging_freq = 1
min_data_in_leaf = 20
seed = 42
```
### Tuned Parameters

```python
learning_rate = 0.5
num_leaves = 31
num_boost_round = 2000
early_stopping_rounds = 100
feature_fraction = 0.8
bagging_fraction = 0.8
bagging_freq = 1
min_data_in_leaf = 100
scale_pos_weight = 50.0
```

---

## 📅 Freshness Analysis (Newly Discovered URLs)

We analyze the proportion of URLs with:

```
2026-04-09 > first_seen >= 2026-04-01
```

### Result

* Count: **5,968,583**
* Total: **32,000,000**
* Ratio: **18.65%**

---

## 🧠 Key Insights

---

### 1. Clear Improvement on Random Coverage

Compared with the Batch3 baseline, the tuned model improves **Random Coverage** from **18.09% → 22.76%**.

- Absolute gain: **+4.67 percentage points**
- Relative gain: **+25.8%**
- URL gain: **+368**

This suggests that the tuned model is able to cover more URLs in the random evaluation set under the same 32M index budget.

---

### 2. Head Coverage Remains Stable with Slight Improvement

For **Head Coverage**, the tuned model improves from **18.39% → 18.45%**.

- Absolute gain: **+0.06 percentage points**
- Relative gain: **+0.03%**
- URL gain: **+40**

Although the gain is small, the result indicates that the tuned model maintains head-query performance while improving random-query coverage.

---

### 3. Better General Coverage Without Sacrificing Head Performance

A key observation is that:

- The tuned model brings a **meaningful gain on Random Coverage**
- While keeping **Head Coverage comparable to the baseline**

This suggests the new configuration improves **broader coverage ability**, rather than only optimizing for a small set of frequent queries.

---

### 4. Effective Incorporation of Newly Discovered URLs

Approximately **18.65%** of the selected 32M URLs satisfy:

```text
2026-04-01 <= first_seen < 2026-04-09
```
This indicates that:

- The model is actively incorporating fresh content
- The improvement is not only from re-ranking old URLs
- New URLs contribute meaningfully to coverage gain

---

### 5. Balanced Freshness Strategy

The freshness ratio (~18.65%) suggests a balanced strategy:

- Not overly biased toward recent URLs
- Preserves historically valuable URLs
- Allows high-potential new URLs into the index

This balance is desirable for production systems.

---

### 6. Tuned Parameters Likely Increase Recall

The tuned configuration introduces:

```python
learning_rate = 0.5
num_boost_round = 2000
scale_pos_weight = 50.0
```
These changes likely:

- amplify positive signals
- improve recall
- contribute to higher Random Coverage

---

### 7. Stronger Regularization May Help Stabilize Selection

The tuned model also increases regularization:

- `min_data_in_leaf`: 20 → 100
- `num_leaves`: 63 → 31

This suggests:

- less overfitting to small patterns
- better generalization

---

### 8. Remaining Risk: Potential Domain Bias

Although coverage improves, further analysis is needed to check:

- domain concentration
- diversity balance

Future validation should include:

- top-domain frequency comparison
- unique domain count
- top-k domain concentration
- overlap with previous 32M selections

---

## 🌐 Domain Distribution Analysis

### 📊 Unique Domain Comparison

| Metric | Batch3 (Old) | Tuned Model (New) |
|-------|-------------:|-----------------:|
| Unique Domains | 7,965 | 528,707 |

---

### 🧠 Insight

The tuned model dramatically increases domain diversity:

- From **7,965 → 528,707 domains**
- Over **66× increase in unique domains**

This indicates that the new model:
- significantly improves **coverage across different websites**
- reduces reliance on a small set of dominant domains
- expands index diversity at a large scale

---

## 🏆 Top Domain Comparison

### Old Model (Batch3)
- Heavy concentration on a few platforms:
  - `github.com`
  - `en.wikipedia.org`
  - `twitter.com`
  - `facebook.com`
- Many domains contribute **hundreds of thousands to millions of URLs**

### Tuned Model
- Still includes major platforms:
  - `en.wikipedia.org`
  - `youtube.com`
  - `instagram.com`
- But distribution is **more spread across multiple domains**
- Wikipedia ecosystem becomes more prominent (multi-language)

---

### 🧠 Insight

- Old model shows **strong concentration bias**
- Tuned model:
  - maintains high-value domains
  - but **distributes selection across many more domains**
- This suggests improved **generalization and diversity**

---

## 📅 Temporal Distribution Analysis

### Old Model (Batch3)

- Dominated by:
  - **2026-03 (~50%)**
  - **2026-02 (~30%)**
- Almost no very recent data

### Tuned Model

- Includes:
  - **2026-04: 5,968,583 URLs (~18.65%)**
  - strong presence of recent data
- Still keeps:
  - large portion from 2026-03 and earlier

---

### 🧠 Insight

The tuned model introduces **recently discovered URLs** while maintaining historical data:

- Adds **freshness without sacrificing stability**
- Shows ability to **integrate newly discovered URLs effectively**
- Aligns with production needs:
  - freshness + consistency

---

## 📈 Score Distribution Observation

### Old Model
- Score range:
  - ~0.20 → 0.42
- Narrow range with relatively low confidence scores

### Tuned Model
- Score range:
  - ~0.99967 → 0.99978
- Extremely high and concentrated scores

---

### 🧠 Insight

- Tuned model produces **high-confidence predictions**
- Likely due to:
  - higher learning rate
  - strong positive weighting (`scale_pos_weight`)
- Suggests:
  - model strongly separates positive vs negative samples
- However:
  - may require further validation to ensure calibration is reasonable

---

## 🔥 Overall Interpretation

The tuned model improves index selection along **three major dimensions**:

### 1. Coverage
- Significant improvement in Random Coverage (+25.8%)

### 2. Diversity
- Massive increase in domain diversity (7,965 → 528,707)

### 3. Freshness
- Incorporates **18.65% newly discovered URLs**

---

## ⚠️ Potential Risks

Despite strong improvements, the following should be monitored:

- **Score saturation**
  - scores are extremely close to 1.0
- **domain distribution skew**
  - Wikipedia ecosystem becomes dominant
- **model calibration**
  - high confidence may not always reflect true ranking quality

---

## ✅ Conclusion (Extended)

Compared to the Batch3 baseline, the tuned model:

- improves overall coverage (especially random queries)
- significantly increases domain diversity
- effectively incorporates new discovered URLs
- maintains head query performance
